resolver/ecs_resolver.go

```diff
@@ -73,25 +75,24 @@ func (r *ECSResolver) Resolve(ctx context.Context, request *model.Request) (*mod
 
 // setSubnet appends the subnet information to the request as EDNS0 option
 // if the client IP is IPv4 or IPv6 and the corresponding mask is set in the configuration
-func (r *ECSResolver) setSubnet(request *model.Request) {
-	e := new(dns.EDNS0_SUBNET)
-	e.Code = dns.EDNS0SUBNET
-	e.SourceScope = ecsSourceScope
-
-	if ip := request.ClientIP.To4(); ip != nil && r.cfg.IPv4Mask > 0 {
-		mip, err := maskIP(ip, r.cfg.IPv4Mask)
-		if err == nil {
-			e.Family = ecsFamilyIPv4
-			e.SourceNetmask = uint8(r.cfg.IPv4Mask)
-			e.Address = mip
+func (r *ECSResolver) setSubnet(so *dns.EDNS0_SUBNET, request *model.Request) {
+	var subIP net.IP
+	if so != nil && r.cfg.Forward && so.Address != nil {
+		subIP = so.Address
+	} else {
+		subIP = request.ClientIP
+	}
+
+	request.Log.Debugf("set edns0 subnet option: %s", subIP)
+
+	if ip := subIP.To4(); ip != nil && r.cfg.IPv4Mask > 0 {
+		if mip, err := maskIP(ip, r.cfg.IPv4Mask); err == nil {
+			e := getEdnsSubnetOption(mip, ecsFamilyIPv4, r.cfg.IPv4Mask)
```

Comment:
Want to avoid `get` meaning "create":
```suggestion
			e := newEdnsSubnetOption(mip, ecsFamilyIPv4, r.cfg.IPv4Mask)
```